In [1]:
#ignore warnings
import warnings
warnings.filterwarnings("ignore")

#import required Libraries
import pandas as pd
import numpy as np

import os 
pd.set_option('display.max_colwidth', -1)
import requests
import io

# Import the stats library
from scipy import stats

import statsmodels.api as sm

# These are the plotting libraries we'll use:
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.mlab as mlab
plt.style.use("classic")
import seaborn as sns

#Machine learning
from sklearn import linear_model
from sklearn.cross_validation import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.metrics import accuracy_score


# Command for plots to appear in the iPython Notebook
%matplotlib inline

/Users/Shravan/anaconda/lib/python3.6/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [7]:
modelingdata = pd.read_csv('modelingdata_sk.csv')

In [8]:
modelingdata.shape

(2579, 268)

In [9]:
modelingdata.head()

avg_age_media_collection  Read Score  Avg_State_Size_EnglishII  \
0  2000.0                    56.0        18.0                       
1  2000.0                    56.0        18.0                       
2  2000.0                    56.0        18.0                       
3  2001.0                    56.0        18.0                       
4  2000.0                    56.0        18.0                       

   lea_cte_courses  szip_ad  st_wap_num  sat_participation_pct  \
0  0.700774         28115    111757      0.92                    
1  0.700774         27560    111757      0.45                    
2  0.700774         27253    111757      0.43                    
3  0.700774         27253    111757      0.45                    
4  0.700774         27244    111757      0.45                    

   st_total_specialized_courses  Avg_Dist_Size_04  st_supplies_expense_pct  \
0  0.964635                      21.0              0.075                     
1  0.964635                      21.0              0.075                     
2  0.964635                      21.0              0.075                     
3  0.964635                      21.0              0.075                     
4  0.964635                      21.0              0.075                     

          ...                 New_ind_Txt  st_avg_daily_attend_pct  \
0         ...           School Is Not New  0.958                     
1         ...           School Is Not New  0.957                     
2         ...           School Is Not New  0.951                     
3         ...           School Is Not New  0.952                     
4         ...           School Is Not New  0.952                     

   lea_benefits_expense_pct  advance_dgr_pct  Avg_State_Size_00  \
0  0.210                     0.268            19.0                
1  0.210                     0.268            19.0                
2  0.212                     0.143            19.0                
3  0.212                     0.194            19.0                
4  0.212                     0.529            19.0                

   English II Score  Avg_Class_Size_01  st_not_highqual_class_lp_pct  \
0  60.0              18.0               0.024                          
1  60.0              19.0               0.024                          
2  60.0              19.0               0.024                          
3  60.0              19.0               0.024                          
4  60.0              18.0               0.024                          

   Overall_Achievement_Score_bins  SchoolWeightedScore  
0  55-69                           173.07               
1  55-69                           173.07               
2  55-69                           173.07               
3  55-69                           173.07               
4  55-69                           173.07               

[5 rows x 268 columns]

In [10]:
modelingdata.Overall_Achievement_Score_bins.head()

0    55-69
1    55-69
2    55-69
3    55-69
4    55-69
Name: Overall_Achievement_Score_bins, dtype: object

In [11]:
modelingdata = pd.read_pickle('modelingdata.pickle')

In [12]:
modelingdata.shape

(2579, 268)

In [13]:
modelingdata.head()

avg_age_media_collection  Read Score  Avg_State_Size_EnglishII  \
0  2000.0                    56.0        18.0                       
1  2000.0                    56.0        18.0                       
2  2000.0                    56.0        18.0                       
3  2001.0                    56.0        18.0                       
4  2000.0                    56.0        18.0                       

   lea_cte_courses  szip_ad  st_wap_num  sat_participation_pct  \
0  0.700774         28115    111757      0.92                    
1  0.700774         27560    111757      0.45                    
2  0.700774         27253    111757      0.43                    
3  0.700774         27253    111757      0.45                    
4  0.700774         27244    111757      0.45                    

   st_total_specialized_courses  Avg_Dist_Size_04  st_supplies_expense_pct  \
0  0.964635                      21.0              0.075                     
1  0.964635                      21.0              0.075                     
2  0.964635                      21.0              0.075                     
3  0.964635                      21.0              0.075                     
4  0.964635                      21.0              0.075                     

          ...                 New_ind_Txt  st_avg_daily_attend_pct  \
0         ...           School Is Not New  0.958                     
1         ...           School Is Not New  0.957                     
2         ...           School Is Not New  0.951                     
3         ...           School Is Not New  0.952                     
4         ...           School Is Not New  0.952                     

   lea_benefits_expense_pct  advance_dgr_pct  Avg_State_Size_00  \
0  0.210                     0.268            19.0                
1  0.210                     0.268            19.0                
2  0.212                     0.143            19.0                
3  0.212                     0.194            19.0                
4  0.212                     0.529            19.0                

   English II Score  Avg_Class_Size_01  st_not_highqual_class_lp_pct  \
0  60.0              18.0               0.024                          
1  60.0              19.0               0.024                          
2  60.0              19.0               0.024                          
3  60.0              19.0               0.024                          
4  60.0              18.0               0.024                          

   Overall_Achievement_Score_bins  SchoolWeightedScore  
0  55-69                           173.07               
1  55-69                           173.07               
2  55-69                           173.07               
3  55-69                           173.07               
4  55-69                           173.07               

[5 rows x 268 columns]

In [15]:
modelingdata.Overall_Achievement_Score_bins.head()

0    55-69
1    55-69
2    55-69
3    55-69
4    55-69
Name: Overall_Achievement_Score_bins, dtype: category
Categories (5, object): [<40 < 40-54 < 55-69 < 70-84 < 85-100]

In [22]:
modelingdata.loc[:, modelingdata.dtypes == bool]

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]

[2579 rows x 0 columns]

In [24]:
modelingdata.summer_program_flg.dtype

dtype('int64')

In [31]:
flags = modelingdata.columns[modelingdata.columns.str.contains('flg')]
print(list(flags))

['stem_flg', 'clp_ind_flg', 'title1_type_flg', 'summer_program_flg', 'focus_clp_flg', 'no_data_spg_flg', 'asm_no_spg_flg']


In [36]:
allFlags = modelingdata.loc[:, list(flags)]
allFlags.head()

stem_flg  clp_ind_flg  title1_type_flg  summer_program_flg  focus_clp_flg  \
0  0         0            0                0                   0               
1  0         0            0                0                   0               
2  0         0            0                0                   0               
3  0         0            0                0                   0               
4  0         0            0                0                   0               

   no_data_spg_flg  asm_no_spg_flg  
0  0                0               
1  0                0               
2  0                0               
3  0                0               
4  0                0

In [37]:
allFlags.summer_program_flg.value_counts()

 0    1702
-1    877 
Name: summer_program_flg, dtype: int64

In [38]:
#Create data = x
df_X = modelingdata.copy()
#create target = Y
df_y = df_X.summer_program_flg
print('X shape before ')
print(df_X.shape)
#drop the response column from the x dataset
df_X.pop('summer_program_flg')
print('X shape after ')
print(df_X.shape)
print('Y shape after ')
print(df_y.shape)

X shape before 
(2579, 268)
X shape after 
(2579, 267)
Y shape after 
(2579,)


In [39]:
def dummy_Create(df):
    schoolDataFinal = df.copy()
    SchoolDataCat = schoolDataFinal.loc[:,(schoolDataFinal.dtypes == bool)| (schoolDataFinal.dtypes == object) |(schoolDataFinal.dtypes == 'category')]#.columnscategory
    # drop the two street address columns
    try: 
        SchoolDataCat.pop('street_ad')
        SchoolDataCat.pop('scity_ad')
    except Exception as e:
        print(e.args)
    print(SchoolDataCat.columns)
    cat_vars=SchoolDataCat.columns
    for var in cat_vars:
        cat_list='var'+'_'+var
        cat_list = pd.get_dummies(schoolDataFinal[var], prefix=var)
        df=pd.concat([df,cat_list], axis=1)
    data=df
    cat_vars=SchoolDataCat.columns
    data_vars=data.columns.values.tolist()
    
    to_keep=[i for i in data_vars if i not in cat_vars]

    ## Final data columns
    data_final=data[to_keep]
    try:
        data_final.pop('street_ad')
        data_final.pop('scity_ad')
    except Exception as e:
        print(e.args)
    data_final.columns.values
    data_final_vars=data_final.columns.values.tolist()
    #return(SchoolDataCat)
    y=['summer_program_flg']
    X=[i for i in data_final_vars if i not in y]
 
    list(data_final_vars)
    print(y)
    print(X)
    return(data_final.abs())
    
df_X = dummy_Create(df_X)

Index(['grades_1_to_1_access', 'type_cd_txt', 'Reading SPG Grade', 'Byod',
       'Grad_project_status', 'esea_status', 'school_type_txt',
       'Math SPG Grade', 'Closed_Ind_Text', 'category_cd', 'state_ad',
       '1_to_1_access', 'grades_BYOD', 'State Gap Compared',
       'EVAAS Growth Status', 'calendar_only_txt', 'SPG Grade', 'New_ind_Txt',
       'Overall_Achievement_Score_bins'],
      dtype='object')
['summer_program_flg']
['avg_age_media_collection', 'Read Score', 'Avg_State_Size_EnglishII', 'lea_cte_courses', 'szip_ad', 'st_wap_num', 'sat_participation_pct', 'st_total_specialized_courses', 'Avg_Dist_Size_04', 'st_supplies_expense_pct', 'RETAINED_School_Pct', 'Avg_Class_Size_07', 'services_expense_pct', 'st_long_susp_per_c_num', 'Avg_Class_Size_04', 'st_prinyrs_0thru3_pct', 'st_emer_prov_teach_pct', 'other_expense_pct', 'Avg_Class_Size_00', 'GCE_ALL_School_Pct', 'st_wap_per_classroom', 'lea_sat_participation_pct', 'books_per_student', 'Avg_Class_Size_Biology', 'GCE_RPF_Schoo

In [40]:
print(df_X.shape)
print(df_y.shape)

(2579, 397)
(2579,)


In [41]:
df_y.value_counts()

 0    1702
-1    877 
Name: summer_program_flg, dtype: int64

## How do I choose which model to use for my supervised learning task ?

Our classification task is to predict `summer_program_flg` of an unknown school profile. We tried using the following classification models:
* Logistic Regression (default parameters)
* Logistic Regression (with model tuning)
* SVM (default parameters)
* SVM (with model tuning)

We received 4 different sets of predictions. 
The goal of supervised learning is to build a model that generalizes to out-of-sample data, that is, what we really need is a procedure that allows us to estimate how well a given model is likely to perform on out-of-sample data. This is known as model evaluation procedure. If we can estimate the likely performance of our 4 models, then we can use that performance estimate to choose between the models.

We will first explore the simplest of model evaluation procedures: `Train and Test on the entire dataset`

### Evaluation procedure 1: Train and Test on the entrire dataset

In [42]:
# import the class
from sklearn.linear_model import LogisticRegression

# instantiate the model (using default parameters)
logreg = LogisticRegression()

# take a look at the default model tuning parameters
logreg

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [43]:
# fit the model with data (model training)
# In this step, the model learns the relationship b/w features in df_X and the reponse df_y.
logreg.fit(df_X, df_y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [44]:
# predict the response for observations in X
y_pred = logreg.predict(df_X)

In [45]:
# check the length of y_pred
len(y_pred)

2579

Now we need a numerical way to evaluate how well our model performed. The most obvious choice would be `Classification Accuracy` which is the proportion of correct predictions. This is our first evaluation metric.

### Compute classification accuracy

We will calculate the classification accuracy using the `metrics` module of scikit-learn:
* First we import the metrics module
* Second we invoke the *accuracy_score* **function**

In [46]:
# compute classification accuracy for our logistic regression model
from sklearn import metrics
metrics.accuracy_score(df_y, y_pred)

0.70492438929817758

In [51]:
print(np.array(df_y[0:100]))
print(y_pred[0:100])

[ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
  0  0  0  0  0  0  0  0  0  0  0  0  0 -1  0  0  0  0  0  0  0  0  0  0  0
  0  0 -1 -1 -1  0 -1 -1 -1 -1 -1 -1 -1  0  0  0  0  0  0 -1 -1 -1 -1 -1 -1
 -1  0  0  0  0  0  0 -1  0 -1  0 -1  0  0  0  0 -1  0 -1 -1 -1 -1 -1 -1 -1]
[ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0 -1 -1
 -1 -1 -1  0 -1 -1 -1 -1 -1 -1 -1 -1 -1  0  0  0  0  0 -1 -1 -1 -1 -1 -1 -1
 -1  0 -1 -1 -1 -1  0  0  0 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1]


That means, it compared 2579 true responses with the corresponding 2579 predicted responses and calculated that 70% of our predictions were correct. This is known as **training accuracy** because we are testing the model on the same data that we used to train the model.

### Evaluation procedure 2: Train/Test split 80-20

In [52]:
# print shapes of df_X and df_y
print(df_X.shape)
print(df_y.shape)

(2579, 397)
(2579,)


To split the data into training and testing sets, we are going to use sckit-learn's built-in, `train_test_split` function.

In [53]:
# STEP 1: split df_X and df_y into training and testing sets
from sklearn.model_selection import train_test_split
df_X_train, df_X_test, df_y_train, df_y_test = train_test_split(df_X, df_y, test_size=0.2, random_state=4)

In [54]:
print(df_X_train.shape)
print(df_X_test.shape)

(2063, 397)
(516, 397)


In [55]:
print(df_y_train.shape)
print(df_y_test.shape)

(2063,)
(516,)


In [56]:
# STEP 2: train the model on the training set
logreg = LogisticRegression()
logreg.fit(df_X_train, df_y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [57]:
# STEP 3: make predictions on the testing set
y_pred = logreg.predict(df_X_test)

In [58]:
# Compute the prediction accuracy
print(metrics.accuracy_score(df_y_test, y_pred))

0.72480620155


### Evaluation procedure 3: Use k-fold Cross-validation

**Goal**: Use cross-validation to find the classification accuracy of default Logistic Regression model.

Our primary **function** for cross-validation in scikit-learn will be `cross_val_score` which we will import from the `sklearn.cross_validation` module.

In [59]:
from sklearn.cross_validation import cross_val_score

In [60]:
# Create the model object
logreg = LogisticRegression()

We will now use the cross_val_score function, giving this function 5 parameters:
* **first parameter**: model object logreg
* **second parameter**: df_X - our feature matrix, it is important to note that we don't pass df_X_train but the entire dataframe. 
* **third parameter**: df_y - our response vector. 
* **fourth parameter**: cv - 10 fold cross validation.
* **fifth parameter**: scoring='accuracy' - which means we want to use classification accuracy as the evaluation metric. 

In [61]:
scores = cross_val_score(logreg, df_X, df_y, cv=10, scoring='accuracy')
print(scores)

[ 0.57142857  0.66023166  0.63565891  0.69767442  0.75581395  0.71705426
  0.53875969  0.73540856  0.75486381  0.6848249 ]


In [62]:
# take the average of the scores
print(scores.mean())

0.675171874823


## Cross-validation example: parameter tuning

**Goal**: Select the best tuning parameters (hyperparameters) for LogisticRegression on this dataset.

In other words, we want to choose the tuning parameters for Logistic Regression that best generalizes to the out-of-sample data. We will be using the C parameter to tune our model.

In [64]:
# search for an optimal value of C for LogisticRegression
c_vals = [1.0, 2.0, 3.0, 0.1, 0.5, 0.01, 0.05, 10, 20, 30, 40, 100]
c_scores = []
for c in c_vals:
    logreg = LogisticRegression(C=c)
    scores = cross_val_score(logreg, df_X, df_y, cv=10, scoring='accuracy')
    c_scores.append(scores.mean())
print(c_scores)

[0.67517187482349006, 0.66973503099645537, 0.66395094954048894, 0.66894781850894414, 0.66548488713903609, 0.66275071107665451, 0.66781201163157555, 0.6693593945562093, 0.66936535731667401, 0.66199799991312069, 0.66701729328640924, 0.66237962240586445]


### More efficient parameter tuning using `GridSearchCV`